### DISCUSSION TITLE
threshold for fp.plot_barcode().

### DISCUSSION THREAD

**Discussion by tjrong123** [chemosim-lab/ProLIF#171] [Q&A] [answered]
i checked the documentation there is no option to add threshold of interaction in the plot_barcode. Is there any method that can decide the threshold for the plot_barcode?

**Comment by cbouy** [comment, marked as answer]
There's no such option, the point of the barcode plot is to show all the individual interactions in every frame.

But from the dataframe you can calculate the mean and then apply a threshold yourself, and then make a regular barplot with matplotlib or any other plotting library.

In the simplest case that would be something like

```python
df = fp.to_dataframe()
dfavg = df.mean()
dfavg = dfavg.loc[dfavg > 0.3].to_frame()
dfavg.plot.bar()
```
(I haven't actually tested that code but it should be a good starting point)

**Comment by Noah603** [comment]
Hello, 
I would like to give my 2cents and say that I would really appriciate a threshold option! 
My Barcode plot (with this function) is horrible to look at, due to the fact that I have a penta-sachharide and 5000 frames. 
With the bar plot you gave, one loses information about when the binding occurs. For example if certain bindings only occur when another binding is happening, it will not be displayed. I will try to create a barcode plot with the df and if I found somthing that works, I will post it here.

**Reply by cbouy** [reply]
There's actually another way to do this shown in issue #184 [here](https://github.com/chemosim-lab/ProLIF/issues/184), I must have been tired when I wrote this answer!

**Reply by Noah603** [reply]
Thank you very very much! Works like a charm :)

On another note, is there a possiblity of showing the atoms which are interacting within the plot? So Instead of Lig-Protein_residue --> LigH5-Protein_Residue_O2 ? The information must be there right?

**Reply by cbouy** [reply]
Combining the answer in #180 with the above and a bit of magic:

```python
import pandas as pd
from prolif.plotting.barcode import Barcode

# extract atom names from molecules and metadata indices
ligand_mol = plf.Molecule.from_mda(ligand_selection)
protein_mol = plf.Molecule.from_mda(protein_selection)

data = []
index = []
for i, frame_ifp in fp.ifp.items():
    index.append(i)
    frame_data = {}
    for (ligres, protres), residue_ifp in frame_ifp.items():
        for int_name, metadata_tuple in residue_ifp.items():
            for metadata in metadata_tuple:
                # extract atom name using atom index in metadata
                prot_atoms = ",".join(
                    [
                        protein_mol.GetAtomWithIdx(x).GetMonomerInfo().GetName()
                        for x in metadata["parent_indices"]["protein"]
                    ]
                )
                lig_atoms = ",".join(
                    [
                        ligand_mol.GetAtomWithIdx(x).GetMonomerInfo().GetName()
                        for x in metadata["parent_indices"]["ligand"]
                    ]
                )
                # set new column name based on ligand and protein atom names
                frame_data[
                    (f"{ligres}|{lig_atoms}", f"{protres}|{prot_atoms}", int_name)
                ] = 1
    data.append(frame_data)

# create dataframe
df = pd.DataFrame(data, index=pd.Index(index, name="Frame"))
df.columns = pd.MultiIndex.from_tuples(
    df.columns, names=["ligand", "protein", "interaction"]
)
df = df.fillna(0).astype(int)

# filter
threshold = 0.1
above_threshold_mask = df.mean() > threshold
df = df.loc[:, above_threshold_mask]

# sort by residues
df = df.sort_index(
    axis=1,
    level="protein",
    key=lambda index: index.map(lambda i: plf.ResidueId.from_string(i.split("|")[0])),
)

# plot
barcode = Barcode(df)
barcode.display(figsize=(12, 10))
```
The rows for pi-stacking interactions aren't looking great though, but it works!

**Reply by Noah603** [reply]
Thank you very very much!!! It creates a nice plot, and on the first two looks the data makes sense!
